### PR TITLE
sftp copy implementation

### PR DESCRIPTION
**Pull request by pcadabam**
Change details for copy design - https://docs.google.com/document/d/1_W2gKxfYneJdwQSwsSvZGIGGgBJW1QptwexQxJLV3Kk/edit

Also see example data flow diagram - https://drive.google.com/open?id=0B2c5NkTp5JlATEdHeVRDREl1d28

The PR includes code change to support copying of files from sftp to hdfs. But works with any FileSystem. The code is tested in gobblin standalone mode. More unit tests in progress.

1. Add a new Source to get all copyable files on 'any' FileSystem
2. Add a new FileSystem implementation for SFTP servers
3. Add Extractor to extract InputStreams.
4. Converter to convert inputStreams (DecryptConverter, UnArchiveConverter) 
5. New Writer to write inputStreams. Also added a new Writer to un tar stream and write
6. Some modifications to BaseDataPublisher to work for copy use case
7. Several changes on the DatasetFinder side to leverage DatasetFinder from retention code.

Required changes this in the design but not in this PR
1. Support of renaming in the publisher
2. Unit tests

@chavdar please take a look.

### CODE REVIEW

**Comment by chavdar** [comment]
@liyinan926 @tuGithub  can you also have a look?

**Comment by chavdar** [comment]
gobblin-test/.DS_Store should not be checked in.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/ArchivedInputStreamDataWriter.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            out.close();
+            btoRead = null;
+          }
+        }
+      }
+    } finally {
+      tarIn.close();
+    }
+
+    // TODO add support for setting permission before commit.
```
Comment (chavdar) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/ArchivedInputStreamDataWriter.java]: Let's implement this before committing.

**Comment by pcadabam** [comment]
Closing this pr. Will be splitting this pr into 3 prs 
1. Dataset
2. Copyable dataset
3. sftp

**Pull request closed without merging**